src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs

```diff
@@ -565,6 +564,17 @@ internal void Configure()
                 {
                     InitializeConstructorParameters(GetParameterInfoValues(), sourceGenMode: Options.SerializerContext != null);
                 }
+
+                // Check for misuse of required properties
+                // Ideally this check should be done during property.EnsureConfigured but at that point we have no knowledge about parametrized constructors
+                // and we cannot check if property is read-only correctly
+                foreach (KeyValuePair<string, JsonPropertyInfo> jsonPropertyInfoKv in PropertyCache.List)
+                {
+                    JsonPropertyInfo jsonPropertyInfo = jsonPropertyInfoKv.Value;
+                    jsonPropertyInfo.CheckRequiredPropertyIsMisused();
```

Comment:
Then this check could be moved inside JsonPropertyInfo.Configure with all the other propertyInfo-level validations.

Reply:
as I mentioned here: https://github.com/dotnet/runtime/pull/78152#discussion_r1019019816 unfortunately that can't be move  (at least not in the current state of the code), the ordering is following and each step depends on the previous:
- we initialize properties
- we initialize ctor parameters
- we check required properties used correctly

if I moved last step to first one we wouldn't know about which property is constructor parameter and be back to original bug